primitives/arithmetic/src/traits.rs

```diff
@@ -302,3 +307,519 @@ pub trait SaturatedConversion {
 	}
 }
 impl<T: Sized> SaturatedConversion for T {}
+
+/// Arithmetic operations with safe error handling.
+///
+/// This module provide a readable way to do safe arithmetics, turning this:
+///
+/// ```
+/// # use sp_arithmetic::{traits::EnsureSub, ArithmeticError};
+/// # fn foo() -> Result<(), ArithmeticError> {
+/// # let mut my_value: i32 = 1;
+/// # let other_value: i32 = 1;
+/// my_value = my_value.checked_sub(other_value).ok_or(ArithmeticError::Overflow)?;
+/// # Ok(())
+/// # }
+/// ```
+///
+/// into this:
+///
+/// ```
+/// # use sp_arithmetic::{traits::EnsureSubAssign, ArithmeticError};
+/// # fn foo() -> Result<(), ArithmeticError> {
+/// # let mut my_value: i32 = 1;
+/// # let other_value: i32 = 1;
+/// my_value.ensure_sub_assign(other_value)?;
+/// # Ok(())
+/// # }
+/// ```
+///
+/// choosing the correct [`ArithmeticError`] it should return in case of fail.
+///
+/// The *EnsureOps* family functions follows the same behavior as *CheckedOps* but
+/// returning an [`ArithmeticError`] instead of `None`.
+///
+/// [`ArithmeticError`]: crate::ArithmeticError
+mod ensure {
+	use super::{CheckedAdd, CheckedDiv, CheckedMul, CheckedSub, Zero};
+	use crate::{ArithmeticError, FixedPointNumber, FixedPointOperand};
+
+	/// Performs addition that returns `ArithmeticError` instead of wrapping around on overflow.
+	pub trait EnsureAdd: CheckedAdd + PartialOrd + Zero + Copy {
```

Comment:
We could maybe get rid of the `Copy`, but it is only used in the error case, so better dont waste time on it.

Reply:
Ok, the ensure_ops_assign methods also makes uses of that copy to return the value (in case we leave them as they are)